### ISSUE TITLE
x265 Does not inherit affinity from StaxRip

### ISSUE THREAD

**Issue by JonathonSunshine** [staxrip/staxrip#194]
**Describe the bug**
If i launch Staxrip and assign it affinity to (for instance), half my cores, and encode with the **x264** codec, with threads set to "0", it will detect however many logical cores are assigned to Staxrip and will not access the other cores at all. On Windows TaskManager, you will see certain cores totally idle.

**Expected behavior**
I expected the HEVC x265 encoder to behave the same way, but it does not. It will detect all the logical cores on the machine and then utilize them. If you set pools to a set number, it will only execute that many threads but they will still go from core to core as they please.

**How to reproduce the issue**
C:\Windows\System32\cmd.exe /c start "StaxRipAlpha" /affinity **3F** "C:\StaxRip\StaxRip.exe"
C:\Windows\System32\cmd.exe /c start "StaxRipAlpha" /affinity **FC0** "C:\StaxRip\StaxRip.exe"

(on a four core/thread system, you would use the hexcodes 3 and C for cores 1+2  and cores 3+4 respectively.

These two commands will start two instances of StaxRip, bound to there own logical processors.
x264 will inherit affinity from Staxrip but x265 will not.

(Clearly this has only become an issue for me since I got a Ryzen 9 and have so many cores I think it is actually a lot faster to do pararelle encodes.

I'm sorry if you feel this was more appropriate as a feature request.

**Comment by Dendraspis** [comment]
You should use x265's own parameters to specify threading options. Limiting a running encoder can cause issues IMHO, because it tries to behave a different way. At least specifying the number of threads/cores should be mandatory.
Have a look at the **Performance** section in **Options**.

> Clearly this has only become an issue for me since I got a Ryzen 9 and have so many cores I think it is actually a lot faster to do pararelle encodes.

Totally depends on your settings. But don't expect too much gain.

**Comment by 44vince44** [comment]
I'd like to confirm that, because unless you are encoding 4k video or more, you don't get advantage of more cores like in the Ryzen 9. So @JonathonSunshine has a point here and if you are encoding 720p or 1080p, you go faster by doing two parrallel encodes of two different files with two different staxrip instances working at the same time.
@Dendraspis I think x265 allow you to select number of cores but not to assign specific cores.
For instance, you can use --pools to tell x265 to use, say, 4 cores. But what we need is to tell x265 to use cores 1,2,3,4 so you can open another instance of x265 and tell it to use cores 5,6,7,8, to ensure the two instances will not use same cores.

**Comment by JonathonSunshine** [comment]
i think that Staxrip just calls the x264 encoder differently perhaps. X264 is executed as a child of the Staxrip process but x265/hevc is launched as it's own process? Also, doing it this way means that Vapoursynth only sees x amounts of cores available.

I dont touch 4k, can't even watch it anyhow, but its nice to be able to dedicate a few cores for whatever else i'm doing as well. I have 12c/24t, ideally I wanna run two jobs parallel, 8 threads each and 8 for whatever i wanna do. Games etc..

@44vince44 yeah, and the benefit of this approach is that i don't have to set the number of threads manually (in x264/avc).    I can set a batch encode list and, x264 will start as many threads as i assign it at launch.

**Comment by 44vince44** [comment]
I think it's not staxrip's fault... It's x265 of creating pools and threding them. I think it's specific to x265 and all encoding frontends will act the same way with x265. I x265 uses a completely different multithreading approach then x264 (this needs to be confirmed by any specialist.. maybe @patman86)

**Comment by Dendraspis** [comment]
Of course this feature would be very nice to get more control over the encoding process, but I think it is questionable if there is a noticeable performance benefit.

@JonathonSunshine 
Setting the process priority to *Below Normal* is enough to play games or whatever while encoding. 

What I wouldn't do is changing the number of assigned cores after x265 started its initialization. This could have a negative impact.

**Comment by 44vince44** [comment]
@JonathonSunshine before continuing further, could you do the following:
open a command promt, and encode a video directly using the x265.exe command of course with the affinity set (without using staxrip). Does the x265.exe use the specified cores ?

**Comment by Patman86** [comment]
Have a look here

[Threading x265](https://x265.readthedocs.io/en/default/threading.html) and here
[Performance Options](https://x265.readthedocs.io/en/default/cli.html?highlight=pools#performance-options)

**Comment by 44vince44** [comment]
VERY useful indeed
;-)))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))

**Comment by JonathonSunshine** [comment]
@Patman86 yeah, i ran the x265 encoder from command line and it launched in the same window and used all cores. I ran the same command, changing x265 to x264 and the encoder launched in its own separate window and ran on the first 6 cores only.

Thanks though, I don't expect you guys to "fix" this.

**Comment by 44vince44** [comment]
Indeed, this proves that the x265 command use an approach that overrides completely any affinity setting. So not a bug, not a fault of staxrip. Please close this thread !

**Comment by Dendraspis** [comment]
Sorry guys, I think this conclusion is wrong. Of course x265 will take all cores, because you tell it to do that via command line.

What is needed is a modification of the x265 command line parameters when the job is being to start. At that moment StaxRip should overwrite the parameters and set the amount of threads to the value StaxRip was started with. Additionally StaxRip should assign the right cores after x265 started.

This should/would work, but it would take some time to realize. But I would bet that Windows is smart enough to make the assignment by itself, so starting 2 encodes limited to 4 cores each (on a 8 core CPU) should be enough.  🤔 
Downside: it's not that flexible due to the fact that the setting is made while creating the job.

But once more: this will only result in better overall performance if a single encode doesn't use all cores!


@44vince44 
Right, StaxRip is completely okay at this point. 😄

**Comment by 44vince44** [comment]
@Dendraspis you're right, but let's keep the answer clear:

1) Johnathon wants to set the affinity on x265, through staxrip (meaning, not only he wants to limit the number of cores, but he wants to assign which core will be working very specifically).
=> this is not possible since starting with affinity options an x265.exe command directly from a cmd window, does not have any effect on which cores will be working. They will all work whatever affinity you select. And this is due to x265 pooling and parallelization of tasks. (test made by @JonathonSunshine, and same command with x264 instead of x265 did have an effect on cores)

2) as per you suggestion, there is a parameters of the x265.exe command that can be set to reduce the number of cores that will be used: --pools parameter. You can insert it as a custom command in the x265 command window in staxrip. Any tampering with other threading parameters can damage performance and even change the encoding results. So consider --pools to be the only parameter you can set, unless you have huge knowledge (and I mean HUUGE) in this area.
This option, itself, will surprise you by not giving you a direct control over the number of threads. it sets the pools, which according to the video's features and other encoding parameters, will decide how many threads will be used. My advice: don't waste your time on that ;-)

3) So, as you @Dendraspis said in a previous post, you can still encode a video and set a lower priority for the x265 process, and create another x265 process using the same lower priority to encode another video, and play a game, the system will handle distribution of instructions across cores.

4) If your concern is only taking more advantage of cores, this will come soon, and @Patman86 promised he'll work on optimizations of the x265 encoder for the Ryzen which has the peculiarity of having too many cores...   

So, the only thing that we can ask from staxrip, is to include latest builds by Patman, which is already the case ;-)

**Comment by 44vince44** [comment]
@stax76 is the chunk encoding that you are working on right now in recent commits, gonna improve x265 performance on these Ryzen processors with many cores ? In that case that would address the issue being discussed here ?

**Comment by stax76** [comment]
I'm not really encoding so people have to find out themselves where it's useful.

This thing took only 4 hours to build because everything needed was already in place, I would have done it before if I knew it's so easy.

**Comment by 44vince44** [comment]
I guess someone with a Ryzen will have to test, @JonathonSunshine ?

**Comment by JonathonSunshine** [comment]
I'm sorry? I read top to bottom, if you want me to test, please link me a built executable?

If you send it real soon, i can test tonight but it's 1.15am here sorry

**Comment by Dendraspis** [comment]
@44vince44 
As far as I can see the overall encoding time should be much higher than running a "normal" encode.  🤔 

@JonathonSunshine 
[StaxRip v2.1.1.3 BETA](https://vf6n4w.am.files.1drv.com/y4myOGqmzBUVL1Hxk8pQH0C3a_vukfMxcelulxWwJLdVzSSgxdWdmCEA0hpn1rXEBfQauWCQyukXdZEG6CppBG61a_prXDi9YcgpTMyjkSmV0DzsOcJqCKbyf0gScJUw-73Rl8kmn-5_tToD7NsKWy0mQ3otfk9bEVzn60ohIBIE9gM5NUFZgbuN_MFLpWFG-VE0IwAvSdv79HfT3cjgeAO1g)

**Comment by 44vince44** [comment]
@JonathonSunshine please test v2.1.1.4 beta (I think 2.1.1.3 doesn't have the chunk stuff in it)
Stax76 just released 2.1.1.4
Check his folder [stax76dropbox](https://www.dropbox.com/sh/4ctl2y928xkak4f/AAADEZj_hFpGQaNOdd3yqcAHa?dl=0)

**Comment by JonathonSunshine** [comment]
ok, quick initital thoughts, its blazingly fast...


![slower 720x480 avc to hevc no resizing or anything](https://user-images.githubusercontent.com/27830114/82232275-c3969000-9971-11ea-8510-6dff51ad628e.jpg)

and on medium, it seems so fast that is suspect something is wrong?

![medium 720x480 avc to hevc no resizing or anything](https://user-images.githubusercontent.com/27830114/82232531-1b34fb80-9972-11ea-9d86-a33ad8bf2d10.jpg)

**Comment by 44vince44** [comment]
@JonathonSunshine wait a second, please in the mean time download v2.1.1.4, I'll be back in 5 minutes to tell you how to test.

remember, the aim here is to have staxrip use as much cpu as possible (we are testing capability of taking most advantage of cpu)

**Comment by JonathonSunshine** [comment]
![image](https://user-images.githubusercontent.com/27830114/82232975-ab734080-9972-11ea-908c-af2bb2f1a35c.png)
ehh scratch that.

the old stable is no slouch either, i didn't realise how much impact QTGMC had on encoding

**Comment by 44vince44** [comment]
@JonathonSunshine is you stay some minutes more with me, we will :
1. test new x265 chunks in ctaxrip
2. I'll give you some easy hints to make QTGMC use all cores (that could really work),

bu one step at a time

**Comment by JonathonSunshine** [comment]
the old stable seems to utilize the cores a lot more evenly. I usually use VapourSynth cause QTGMC seems 5 times faster than AVS.

I HAVE to go to bed, i'm happy to run whatever tests you want, so if you can specify settings or whatnot, i'll do it tomorrow. Night guys, stay safe

**Comment by Dendraspis** [comment]
@44vince44 
Regarding the changelog it was already included.

@JonathonSunshine 
How many *chunks* did you set up?

I wonder what you guys are aiming now. This *chunk thing* would only help, if x265 couldn't use all cores/threads, what it absolutely does!

**Comment by JonathonSunshine** [comment]
@Dendraspis i get it you jelly but 34% is not using "all cores"

**Comment by 44vince44** [comment]
@JonathonSunshine : chunk should have an effect in your case. It's worth to do the test. 

**Test**
when you download 2.1.1.4 you'll need to open the x265 options, go to "Other" then increase chunks (let's say try 4, maybe 8, maybe 16), and do an encoding with no filters at all, only to benchmark the encoder. use q=18, mode = slower. And let's see what happens.

Now if you want the trick to make QTGMC and many AVS filter work better is to add at the end of the filters list, the filter Misc > MTMMode > Prefetch, value 16

**Comment by 44vince44** [comment]
OH YEAH IT DOES :-}}}}}}}}

**Comment by JonathonSunshine** [comment]
@44vince44 set chunks to 4 but it only does two parallelle encodes, it try it again with longer clip.

**Comment by JonathonSunshine** [comment]
the other thread was about 7fps
![image](https://user-images.githubusercontent.com/27830114/82234777-2f2e2c80-9975-11ea-96a1-4b0459dcefbc.png)


Wait, it finished, and then it started another two threads

**Comment by JonathonSunshine** [comment]
![image](https://user-images.githubusercontent.com/27830114/82235289-de6b0380-9975-11ea-99db-b622fbfea95e.png)
 8fps, 5, 5, 5,

**Comment by 44vince44** [comment]
WAIT: did you see the CPU, now it is working !!! you're using 91% cpu, you were using less than 30%.
Now that's a HUGE improvement (91% is a very very good thing)

**Comment by JonathonSunshine** [comment]
10, 5, 5, 5 more or less at 6 threads, i'm sure i can tweak it.
![image](https://user-images.githubusercontent.com/27830114/82235715-70730c00-9976-11ea-8d01-57878bdd5a98.png)

**Comment by JonathonSunshine** [comment]
> 
> 
> WAIT: did you see the CPU, now it is working !!! you're using 91% cpu, you were using less than 30%.
> Now that's a HUGE improvement (91% is a very very good thing)

well yeah, its better than parallel incodes, and it's what i'm gonna do from now on?

buts its also kinda the same thing as paralle encodes.

I'm just not convinced that 96 simultaneous threads is the best way to go. I don't know multithreading works in AVS/StaxRip/VapourSynth either but maybe limit the number of threads there somehow as well.

**Comment by 44vince44** [comment]
I think you're mixing threads and chunks aren't you ?

The best is to use 4 threads ans 4 chunks

**Comment by Patman86** [comment]
i can confirm the improvement by chunks and parallel processes!!!

![grafik](https://user-images.githubusercontent.com/54327252/82236306-32ed8f00-9934-11ea-9e30-c90a747fde4b.png)

THX stax

**Comment by JonathonSunshine** [comment]
each x265 encode said it was creating a pool of 24 threads, so i limit it to 6 threads.

(i couldn't justify the beastly 3950, the price per core starts getting really high

**Comment by 44vince44** [comment]
@JonathonSunshine where are the "threads" you are talking about ? I don't think I' m following

**Comment by JonathonSunshine** [comment]
POOLS
![image](https://user-images.githubusercontent.com/27830114/82236647-bb415380-9977-11ea-882b-9a07bafddbbf.png)

not trying to rain on the parade, its definitely better than encoding 4 videos at thesame time, but is it actually faster? Also, i feel you should either retitle this thread or start a new one cause it still doesn't look like my original issues getting fixed. (ie, have 4 cores that don't get used at all for my browsing etc)

**Comment by JonathonSunshine** [comment]
in my last screenshot of staxrip, it was using 6 pools per encode, in the ones above, it was set to automatic, using 24 pools (each)

**Comment by 44vince44** [comment]
@JonathonSunshine Please do not set number of pools.
Leave it emply.
Just set x265 chunks = 4
            staxrip max parallel processes = 4
This should be the optimal thing (for now).

**Comment by JonathonSunshine** [comment]
i gotta run guys, sorry, i be back in 13 hours but i got a lot of uni shht to do as well so try and set out any tasks you want me to do.

**Comment by 44vince44** [comment]
@JonathonSunshine is it actually faster ? yes it is faster when you are processing videos that are below 720p. For  4K it should be the same

**Comment by 44vince44** [comment]
@JonathonSunshine We're done here!

Thanks @stax76 this is very very good !

**Comment by Dendraspis** [comment]
Has anybody proven the start/end of each chunk?

**Comment by 44vince44** [comment]
what do you mean ? if they match, well yes they seem to, the video id ok, not discontinuities

**Comment by JonathonSunshine** [comment]
@44vince44 but just to confirm, all thats happening here is running four instances of x265 at the same time, it's not a new build of x265 itself?

**Comment by JonathonSunshine** [comment]
"all thats happening" (sorry if that sounds rude)

**Comment by 44vince44** [comment]
No rudeness detected or taken here. Yes it is running instances of x265 at the same time. This is an improvement of x265 since version 2.9 i think. So it's time ago. So it is due to new x265 build optimization

**Comment by Dendraspis** [comment]
@44vince44 
Yes, if they are visually okay. I guess for 480p it should be enough to check that.

>  This is an improvement of x265 since version 2.9 i think.
There is an improvement that makes this solution possible, but the initial extension was not meant to split an encode!

**Comment by 44vince44** [comment]
Benchmarking, a 480p video on my i7, 480p chosen on purpose because i can't get the full usage % on my cpu with that resolution.

- no chunks : cpu % 65-74, encoding time 130s
- using 4 chunks and 4 parallel processes : cpu% at 96 all the time, encoding time = 108s

So 17% better performance.

**Comment by JonathonSunshine** [comment]
unrelated, i set mmmode to prefetch 12, (i have 12 core, not 16 :( , i presume thats how you chose that number.) VapourSynth still WAY faster (assuming it is actually equivalent)

**Comment by 44vince44** [comment]
12? Why do I see 16 logical cpus on you task manager ? 16 logical cpus = 16 threads,
so you need to put prefetch(16).

**Comment by JonathonSunshine** [comment]
> 
> 
> 12? Why do I see 16 logical cpus on you task manager ? 16 logical cpus = 16 threads,
> so you need to put prefetch(16). However 12 is also a good value. The important is that it's not 1 or 2.

that is patmans

is there an equivalent ffor vapoursynth?

**Comment by 44vince44** [comment]
LOL you have 24 logical cores :)))))))))))
physical cores = 12 => logical cores = 24
So you can set prefetch(24)

Just monitor the cpu usage

**Comment by 44vince44** [comment]
> is there an equivalent ffor vapoursynth?

what do you mean

**Comment by JonathonSunshine** [comment]
> 
> 
> > is there an equivalent ffor vapoursynth?
> 
> what do you mean


I feel that 3 simultaneous chunks with 8 threads/pools each is optimal. (based on 8x3=24=ryzen 3900x)
I guess on a 16c/32thread system you would do 8 threads by 4 chunks simultaneously.

SOOOOO... i set 8 pools, 3 chunks simultaneous, i also set total chunks to 24, one always seems to be way faster than the others so this way it spends less time on that one chunk thats left to encode at the end.
 --crf 18 --preset slow --profile main10 --output-depth 10 --pools 8 --sar 10:11 
(i'm getting about 70fps, QTGMC Slower, and a few deblocking filters and color etc.)
![image](https://user-images.githubusercontent.com/27830114/82246684-7f62ba00-9988-11ea-98a9-636fe4b9ec55.png)

**Comment by 44vince44** [comment]
@JonathonSunshine vapoursynth does not require mtmmode setting, because being more recent, it multithreads automatically. And BTW Avisynth plugins that have been recently updated also don't need the prefetch(x) filter.

Now that you have reached 96% usage of your CPU, then you're done !
I think you can close this thread because the issue has been adressed!
Thanks to @stax76

**Comment by Dendraspis** [comment]
Oh boy,  I totally misinterpreted this new feature. I was irritated because of the "chunk feature from x265 since v2.9" mentioned before, but as far I can see now there is no use of x265's chuck feature but the trimming is being applied to the piping app! So it's nearly the same as cutting the source into x parts, encode each part separately with an individual StaxRip instance in parallel and merge them together.

So have fun using it, I'm out.  😄

**Comment by 44vince44** [comment]
maybe I did say something incorrect. this chunk is not the the one introduced in 2.9.
that other is --chunk-start and --chunk-end 
This one is something else (trim of avs2pipemod).... My mistake ... I thought they were related.
sorry for the confusion...

**Comment by Dendraspis** [comment]
@44vince44 
**No**, it was my fault not recognizing it directly.  😄 

*Little sidenote:  This option should be in your mind on upcoming playback and compatibility issues.*

**Comment by JonathonSunshine** [comment]
So like.... I know you guys wanna close this thread and all, and chunks is cool and all but if you can't get x265 to inherit affinity then isn't it more up to you guys to mark it as "won't fix" or ????

Not trying to be rude cause i know i'll be back here someday... tbh, as soon as it's marked

**Comment by 44vince44** [comment]
@JonathonSunshine the test I told you to do and that you did (remember:trying to set affinity for x265 in a command line) showed that x265 does not care about affinity.
So staxrip is doing its job, the affinity is naturally inherited to the command line x265 (or any command line from staxrip). But then x265 does not take it into account, because it launches pools of processing.
So there is **NOTHING** staxrip can do about that ! there is nothing to "fix" with staxrip. It's due to the design of x265.
Besides Stax76 gave us another option by adding the chunk feature.

**Comment by Dendraspis** [comment]
> So there is **NOTHING** staxrip can do about that !

Of course it *could* do something:
> What is needed is a modification of the x265 command line parameters when the job is being started. At that moment StaxRip should overwrite the parameters and set the amount of threads to the value StaxRip was started with. Additionally StaxRip should assign the right cores after x265 started.
> This should/would work, but it would take some time to realize.

But I'm sure this issue is going to be closed as *'too time consuming to realize'* and I wouldn't complain. 🤔  
The best and most save workaround is to set the number of threads/pools by hand while creating/editing the job.

**Comment by stax76** [comment]
I've looked at the help for the start command and the affinity option, it's not really a topic I'm well-informed.

I don't know if there is something that needs to be fixed or improved in staxrip regarding this topic.

**Comment by 44vince44** [comment]
@stax76 :
- Affinity has no effect on the x265 command due to the way x265 is coded and handles threads. So here there is nothing staxrip can do.
- x265 threads can be controlled with the --pools parameter, and you can access it from the x265 options window. Of course it requires knowledge, which is not surprising.

So I guess currently there is nothing to do here

**Comment by JonathonSunshine** [comment]
it's all good. I thought that x265 was being accessed in a different way somehow than x264, it's not happening and I'm ok with that.
